src/parser.ts

```diff
@@ -0,0 +1,749 @@
+				resultG.value,
+			),
+		);
+	});
+
+/**
+ * Similar to `map2`, but for more parsers. See the docs for that function for
+ * more details and examples.
+ */
+const map8 = <A, B, C, D, E, F, G, H, I>(
+	f: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H) => I,
+) => (
+	pa: Parser<A>,
+	pb: Parser<B>,
+	pc: Parser<C>,
+	pd: Parser<D>,
+	pe: Parser<E>,
+	pf: Parser<F>,
+	pg: Parser<G>,
+	ph: Parser<H>,
+): Parser<I> =>
+	parser((a) => {
+		const resultA = pa.run(a);
+
+		if (resultA.kind === ResultKind.Err) {
+			return resultA;
+		}
+
+		const resultB = pb.run(a);
+
+		if (resultB.kind === ResultKind.Err) {
+			return resultB;
+		}
+
+		const resultC = pc.run(a);
+
+		if (resultC.kind === ResultKind.Err) {
+			return resultC;
+		}
+
+		const resultD = pd.run(a);
+
+		if (resultD.kind === ResultKind.Err) {
+			return resultD;
+		}
+
+		const resultE = pe.run(a);
+
+		if (resultE.kind === ResultKind.Err) {
+			return resultE;
+		}
+
+		const resultF = pf.run(a);
+
+		if (resultF.kind === ResultKind.Err) {
+			return resultF;
+		}
+
+		const resultG = pg.run(a);
+
+		if (resultG.kind === ResultKind.Err) {
+			return resultG;
+		}
+
+		const resultH = ph.run(a);
+
+		if (resultH.kind === ResultKind.Err) {
+			return resultH;
+		}
+
+		return ok(
+			f(
+				resultA.value,
+				resultB.value,
+				resultC.value,
+				resultD.value,
+				resultE.value,
+				resultF.value,
+				resultG.value,
+				resultH.value,
+			),
+		);
+	});
+
+/**
+ * Like `map` but applies a function that *also* returns an `Parser`. Then
+ * "unwraps" the result for you so you don't end up with `Parser<Parser<A>>`.
+ * Can be useful for stringing together multiple parsing steps, where each step
+ * relies on the result of the previous one.
+ *
+ * If the first parser fails, the function won't be called and the error will
+ * be returned instead. If the second parser fails the error for that will be
+ * returned.
+ *
+ * @param f The function to apply
+ * @param pa The parser whose result is to be passed to the function
+ * @return {Parser<B>} A new parser, built from `pa` and `f`
+ * @example
+ * type MultipleChoiceAnswer = 'a' | 'b' | 'c';
+ *
+ * const answerValueParser = (value: string): Parser<MultipleChoiceAnswer> =>
+ *   value === 'a' || value === 'b' || value === 'c'
+ *     ? succeed(value)
+ *     : fail("Needed 'a', 'b' or 'c'.");
+ *
+ * const answerParser = pipe(
+ *   fieldParser('answer', parseString),
+ *   andThen(answerValueParser),
+ * );
+ *
+ * const jsonA: unknown = JSON.parse('{ "answer": "a" }');
+ * const resultA = parse(answerParser)(jsonA); // Ok<MultipleChoiceAnswer>, 'a'
+ *
+ * const jsonB: unknown = JSON.parse('{ "answer": "d" }');
+ * const resultB = parse(answerParser)(jsonB); // Err<string>, "Needed 'a', 'b' or 'c'."
+ */
+const andThen = <A, B>(f: (a: A) => Parser<B>) => (pa: Parser<A>): Parser<B> =>
+	parser((a) => resultAndThen<string, A, B>((x) => f(x).run(a))(pa.run(a)));
+
+const oneOf = <A>(parsers: Array<Parser<A>>): Parser<A> =>
```

Comment:
I know we spoke about in IRL but just wanted to add here, it would be good to add a description/examples like you have done with the other functions.